node/src/blockchain/errors/validation_status.rs

```diff
@@ -151,4 +224,97 @@ mod tests {
+        let err = BlockchainErrorKind::AppRpc(AppRpcErrorKind::Internal);
+        let timestamp = UNIX_EPOCH
+            .checked_add(Duration::from_secs(1234567890))
+            .unwrap();
+        let clock = ValidationFailureClockMock::default().now_result(timestamp);
+
+        let result = serde_json::to_string(&PreviousAttempts::new(err, &clock)).unwrap();
+
+        assert_eq!(
+            result,
+            r#"[{"error":{"AppRpc":"Internal"},"firstSeen":{"secs_since_epoch":1234567890,"nanos_since_epoch":0},"attempts":1}]"#
+        );
+    }
+
+    #[test]
+    fn previous_attempts_custom_serialize_seq_initialization_err() {
+        let mock = SerdeSerializerMock::default()
+            .serialize_seq_result(Err(serde_json::Error::custom("lethally acid bobbles")));
+        let err = BlockchainErrorKind::AppRpc(AppRpcErrorKind::Internal);
+        let timestamp = UNIX_EPOCH
+            .checked_add(Duration::from_secs(1234567890))
+            .unwrap();
+        let clock = ValidationFailureClockMock::default().now_result(timestamp);
+
+        let result = PreviousAttempts::new(err, &clock).serialize(mock);
+
+        assert_eq!(result.unwrap_err().to_string(), "lethally acid bobbles");
+    }
+
+    #[test]
+    fn previous_attempts_custom_serialize_seq_element_err() {
+        let mock = SerdeSerializerMock::default()
+            .serialize_seq_result(Ok(SerializeSeqMock::default().serialize_element_result(
+                Err(serde_json::Error::custom("jelly gummies gone off")),
+            )));
+        let err = BlockchainErrorKind::AppRpc(AppRpcErrorKind::Internal);
+        let timestamp = UNIX_EPOCH
+            .checked_add(Duration::from_secs(1234567890))
+            .unwrap();
+        let clock = ValidationFailureClockMock::default().now_result(timestamp);
+
+        let result = PreviousAttempts::new(err, &clock).serialize(mock);
+
+        assert_eq!(result.unwrap_err().to_string(), "jelly gummies gone off");
+    }
+
+    #[test]
+    fn previous_attempts_custom_serialize_end_err() {
+        let mock =
+            SerdeSerializerMock::default().serialize_seq_result(Ok(SerializeSeqMock::default()
+                .serialize_element_result(Ok(()))
+                .end_result(Err(serde_json::Error::custom("funny belly ache")))));
+        let err = BlockchainErrorKind::AppRpc(AppRpcErrorKind::Internal);
+        let timestamp = UNIX_EPOCH
+            .checked_add(Duration::from_secs(1234567890))
+            .unwrap();
+        let clock = ValidationFailureClockMock::default().now_result(timestamp);
+
+        let result = PreviousAttempts::new(err, &clock).serialize(mock);
+
+        assert_eq!(result.unwrap_err().to_string(), "funny belly ache");
+    }
+
+    #[test]
+    fn previous_attempts_custom_deserialize_happy_path() {
+        let str = r#"[{"error":{"AppRpc":"Internal"},"firstSeen":{"secs_since_epoch":1234567890,"nanos_since_epoch":0},"attempts":1}]"#;
```

Comment:
Here too, we're losing the precision here. We also want to know if it's remote or local.